src/dotnet/ShellShimMaker/ShellShimMaker.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.Extensions.EnvironmentAbstractions;
+
+namespace Microsoft.DotNet.ShellShimMaker
+{
+    public class ShellShimMaker
+    {
+        private readonly string _pathToPlaceShim;
+
+        public ShellShimMaker(string pathToPlaceShim)
+        {
+            _pathToPlaceShim =
+                pathToPlaceShim ?? throw new ArgumentNullException(nameof(pathToPlaceShim));
+        }
+
+        public void CreateShim(string packageExecutablePath, string shellCommandName)
+        {
+            var packageExecutable = new FilePath(packageExecutablePath);
+
+            var script = new StringBuilder();
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                script.AppendLine("@echo off");
+                script.AppendLine($"dotnet {packageExecutable.ToQuotedString()} %*");
```

Comment:
How do you intend to deal with different kinds of applications?

Reply:
The code today will error on non dotnet core, however metadata XML are supporting it which is the only thing that is hard to change. In the future, there should be different strategy object to compose this class different type * different os 

All the code associate with it are internal to dotnet.dll, it can be changed easily